### ISSUE TITLE
Jenkins webhook proxy should create pipeline based on HTTP/ params

### ISSUE THREAD

**Issue by clemensutschig** [opendevstack/ods-core#80]
the webhook proxy should take HTTP (get?paramA=B&paramB=C) params on request and transform them into BC pipeline ENV params.

**Comment by michaelsauter** [comment]
See `env` section in https://docs.openshift.com/container-platform/3.9/rest_api/apis-build.openshift.io/v1.BuildConfig.html.

**Comment by stitakis** [comment]
@clemensutschig trying to visualize this use case... why should the webhook proxy accept parameters? How does this differs from current implementation? Thanks!

**Comment by clemensutschig** [comment]
@stitakis  suppose I want to call the webhook from jira /when closing a release (to start the RM process opendevstack/ods-project-quickstarters#62) - at that point its unlikely to have a json bitbucket payload at hand .. but only a few http get params. makes sense?

**Comment by stitakis** [comment]
@clemensutschig I see, this is related to RM and promoting images... I think we should work out first how we want to implement the release process. Triggering the process from Jira, doesn´t sound to me that fits in the context of DevOps. Another detail, to call the webhook from Jira will require to add the webhook trigger_key in Jira and share it with all project members.

**Comment by michaelsauter** [comment]
@stitakis Agreed. After thinking a bit more about this, I would keep this issue open, but not start implementing. It is unclear how things should work. A call from Jira might have a completely different payload, so then we could add the envs there, and not to the get params.

So once we know how the release works, we can figure out what the proxy needs to support, and build that. Because of this, I would remove from 1.1.0. Agreed?

**Comment by stitakis** [comment]
sounds good to me

**Comment by ungerts** [comment]
Tested pipline ENV params. Please consider that we have to extend the triggers in build config (bc) with parameter _allowEnv_. Otherwise parameters are not passed through.
```
triggers:
    - generic:
        allowEnv: true
        secret: supergeheim
      type: Generic
```
Furthermore, I tried to implement a release endpoint PoC in webhook proxy. I think we should discuss which set of parameters is needed to identify the respective build config. In my test I had to pass _branch_, _project key_ and _repo_. To my knowledge, this information is currently being gathered from the bitbucket webhook payload. My payload is as follows (POST request):
```
{
   "branch":"<branch>",
   "project":"<project key>",
   "repo":"<repo name>",
   "env":[
      {
         "name":"<variable name>",
         "value":"<variable value>"
      }
   ]
}
```
Based on the message I could determine the pipline name (by using the existing function _makePipelineName_):
```
project = strings.ToLower(json.Project)
branch = json.Branch
repo = json.Repo
component = strings.Replace(repo, project+"-", "", -1)
pipeline := makePipelineName(project, component, branch)
gitURI := fmt.Sprintf("%s/%s/%s.git", s.RepoBase, project, repo)
```

**Comment by ungerts** [comment]
As discussed with @metmajer:
**Endpoint:** /release
**HTTP Method:** POST
**Request Payload:**
```
{
  "x": "y"
}
```
Example:
```
{
  "VERSION": "v1.0.0",
  "ISSUE": "#80"
}
```
Composite keys should be in snake case and uppercase. That allows us to access the env variables in Jenkins using the pattern env.VARABLE_NAME
**Expected Response Code:** 201 (Created)
Build is only triggered and will not finish within the HTTP call.
**Expected Response:**
```
{
    "status": "Running", //delivered by Openshift API call
    "buildNumber": "43" //delivered by Openshift API call
    "selfLink": "<URL>" // url to fetch build status via webhook proxy (cf. issue #120)
}
```

**Comment by michaelsauter** [comment]
@ungerts I am not sure I get how both your message correlate.

I like the look of the first payload, and this is what I had in mind too. 

The response is already implemented, see #129. It just passes through whatever OpenShift gives back.

**Comment by ungerts** [comment]
Second message. is the result of a short discussion with @metmajer. The idea was to simplify the payload. To conclude...

**First Payload**
```
{
   "branch":"<branch>",
   "project":"<project key>",
   "repo":"<repo name>",
   "env":[
      {
         "name":"<variable name>",
         "value":"<variable value>"
      }
   ]
}
```
**Second Payload**
```
{
  "branch": "<branch to build>",
  "project": "<project key>",
  "repo": "<repo name>",
  "variableA": "valueA",
  "variableB": "valueB"
}
```
_branch_, _repo_, _project_ would be used be used to determine the build config. Everything else will be forwarded to Jenkins (env var).

Maybe Option 1 is the safer option, as it is clear which inputs are expected ( _branch_, _repo_, _project_ ) Opinions?

**Comment by michaelsauter** [comment]
Definitely the first option. The second payload has unknown keys, and is therefore hard to parse.

**Comment by metmajer** [comment]
@michaelsauter @ungerts I'd prefer to have the payload say `params` instead of `env`. The former is generic, while the latter is specific. It's an implementation detail that we'll transform params and arguments into environment variables, but this is a detail that may change over time (who knows).

**Comment by rattermeyer** [comment]
I am with @michaelsauter . Definitively the first option. If params or env doesn't matter in my opinion. neither one is more generic. params is for methods, we do not have methods could someone argue.
So choose one, either one is fine.

**Comment by metmajer** [comment]
@rattermeyer your understanding about params shouldn't end with methods. According to Wikipedia: "A parameter, generally, is any characteristic that can help in defining or classifying a particular system."

**Comment by michaelsauter** [comment]
I guess that whether you prefer `params` or `env` depends how you view the webhook proxy. Is that thing something generic that abstracts what is behind, or is it a simply proxy and mirrors what is behind? If you see it like the former, one would go with `params`, if you see it like the latter, one would go with `env` I believe (even though `env` still feels pretty generic to me). I have always seen the webhook proxy like a simple proxy, and therefore prefer `env`. But in the end I believe this will not make a huge difference. I will pick one option later today with @clemensutschig.

**Comment by metmajer** [comment]
Good hint. Totally agree if we see this merely as a proxy.

**Comment by michaelsauter** [comment]
We selected option 1, with an additional query param called `component`.

**Comment by metmajer** [comment]
@michaelsauter can you explain the need for the extra `component` param? Also, if you adding a query parameter, you're mixing HTTP GET (query params) with POST (all data enclosed in body) capabilities, and create a non-RESTful API.

**Comment by michaelsauter** [comment]
@metmajer Sure. I was about to explain this further in the PR, we just wanted to document it.

The idea of adding a query param came up with #135. The implemented feature there allows to specify an alternative `Jenkinsfile` path. This opens up the possibility to have multiple pipelines for one repository IF we can also specify an alternative pipeline name. Right now, the pipeline name is derived from project, repository and branch name information within the BitBucket payload. As we cannot change the content of the BitBucket payload (body) the only way to trigger multiple pipelines from the same repo is by specifying two webhooks with different query params.

Now if we allow users to specify the "component" (the part that is currently derived from the repo) then we can offer that feature. We thought this feature also makes sense for this "open" endpoint that allows to send a simple payload to the webhook proxy to create/trigger a pipeline. Maybe the thing that wants to trigger a release wants to control the name of the pipeline?

That is why we would offer a query param called `component`. I know this is not the proper way (tm) but it seems to be the most pragmatic one to solve a specific problem. And I've never seen a problem in the real world with doing POST requests while also specifying GET params ...
